File: nestedrecyclerview/src/main/java/com/yang/nestedrecyclerview/ChildRecyclerView.java
package com.yang.nestedrecyclerview;

import android.content.Context;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.ViewParent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

public class ChildRecyclerView extends RecyclerView {
    private static final String TAG = "ChildRecyclerView";

    private FlingHelper mFlingHelper;
    private int mMaxDistance = 0;
    private int mVelocityY = 0;
    private boolean isStartFling;
    private int totalDy;

    private ParentRecyclerView mParentRecyclerView;
    
    public ChildRecyclerView(@NonNull Context context) {
        this(context, null);
    }

    public ChildRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ChildRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mFlingHelper = new FlingHelper(context);
        mMaxDistance = mFlingHelper.getVelocityByDistance(UIUtils.getScreenWidth() * 4);
        setOverScrollMode(OVER_SCROLL_NEVER);
        initScrollListener();
    }

    private void initScrollListener() {
        addOnScrollListener(new OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                Log.i(TAG, "onScrollStateChanged: " + newState);
                if (newState == SCROLL_STATE_IDLE){
                    dispatchParentFling();
                }
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                Log.i(TAG, "onScrolled: " + dy + "    isStartFling:" + isStartFling);
                if (isStartFling){
                    totalDy = 0;
                    isStartFling = false;
                }
                totalDy += dy;
            }
        });
    }

    private void dispatchParentFling() {
        ParentRecyclerView parentRecyclerView = getParentRecyclerView();
        if (parentRecyclerView == null){
            Log.i(TAG, "dispatchParentFling: parentRecyclerView cannot be null");
            return;
        }
        boolean scrollTop = isScrollTop();
        Log.i(TAG, "dispatchParentFling: scrollTop->" + scrollTop + "   mVelocityY->" + mVelocityY);
        if (scrollTop && mVelocityY != 0) {
            double flingDistance = mFlingHelper.getSplineFlingDistance(mVelocityY);
            if (flingDistance > (Math.abs(totalDy))){
                fling(0, -mFlingHelper.getVelocityByDistance(flingDistance + totalDy));
            }
            totalDy = 0;
            mVelocityY = 0;
        }
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        if (ev.getAction() == MotionEvent.ACTION_DOWN){
            mVelocityY = 0;
        }
        return super.dispatchTouchEvent(ev);
    }

    @Override
    public boolean fling(int velocityX, int velocityY) {
        if (!isAttachedToWindow()){
            return false;
        }
        boolean fling = super.fling(velocityX, velocityY);
        Log.i(TAG, "fling: " + fling + "  " + velocityY);
        if (!fling || velocityY >= 0){
            mVelocityY = 0;
        } else {
            isStartFling = true;
            mVelocityY = velocityY;
        }
        return fling;
    }

    /**
     * 是否已经滚动到顶部
     * @return
     */
    public boolean isScrollTop() {
        //RecyclerView.canScrollVertically(-1)的值表示是否能向下滚动，false表示已经滚动到顶部
        return !canScrollVertically(-1);
    }


    private ParentRecyclerView getParentRecyclerView(){
        Log.i(TAG, "getParentRecyclerView: ");
        if (mParentRecyclerView != null){
            return mParentRecyclerView;
        }

        ViewParent parent = getParent();
        while (parent != null){
            if (parent instanceof ParentRecyclerView){
                mParentRecyclerView = (ParentRecyclerView) parent;
                break;
            }
            parent = parent.getParent();
        }
        return mParentRecyclerView;
    }
    
    public void setChildRecyclerViewForParent(){
        ParentRecyclerView parentRecyclerView = getParentRecyclerView();
        if (parentRecyclerView != null){
            Log.i(TAG, "setChildRecyclerViewForParent: ");
            parentRecyclerView.setChildRecyclerView(this);
        }
    }
}

File: ibl_scrollrecyclerview/src/main/java/com/yang/ibl/scrollrecyclerview/library/ChildRecyclerView.kt
package com.yang.ibl.scrollrecyclerview.library

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import androidx.recyclerview.widget.RecyclerView
import java.lang.ref.WeakReference
import kotlin.math.max

class ChildRecyclerView : RecyclerView, OnChildFlingListener {
    constructor(context: Context) : this(context,null)
    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs,0)
    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle)

    /** true 开启滑动冲突处理*/
    var enableConflict = true
    private var draggingY = 0
    private var draggingTime = 0L
    /** 记录上次的parent,避免递归频繁*/
    private var parentView :WeakReference<OnParentFlingListener>? = null

    init {
        addOnScrollListener(object :OnScrollListener(){
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                Log.i(TAG, "onScrollStateChanged: $newState")
                /** 滚动停止且到了顶部,快速滑动事件往上给parent view*/
                if(enableConflict && SCROLL_STATE_IDLE == newState) {
                    induceParentOfChildTopStatus()
                    val step = (System.currentTimeMillis()-draggingTime).toInt()
                    val speed = 1000*draggingY/max(1000,step)
                    parentView?.get()?.onParentFling(speed)
                }else if(SCROLL_STATE_DRAGGING == newState){
                    draggingY = 0
                    draggingTime = System.currentTimeMillis()
                }
            }

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                Log.i(TAG, "onScrolled: $dy")
                if(SCROLL_STATE_DRAGGING == scrollState) {
                    draggingY += dy
                }
            }
        })
    }

    override fun onChildFling(speed: Int) {
        fling(0,speed)
    }

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        Log.i(TAG, "dispatchTouchEvent: $enableConflict")
        if(enableConflict) {
            induceParentOfChildTopStatus()
        }
        return super.dispatchTouchEvent(ev)
    }

    private fun induceParentOfChildTopStatus(){
        /** true child在顶部*/
        findOnParentFlingListener()?.onScrollTop(!canScrollVertically(-1),top)
    }

    private fun findOnParentFlingListener(): OnParentFlingListener? {
        val flingListener = parentView?.get()
        if (flingListener != null){
            return flingListener
        }
        var pv = parent
        while (pv != null) {
            if (pv is OnParentFlingListener) {
                parentView = WeakReference(pv)
                return pv
            }
            pv = pv.parent
        }
        return null
    }

    companion object{
        const val TAG = "ChildRecyclerView"
    }
}